### PR TITLE
Trace search client configuration

### PR DESCRIPTION
**Pull request by majorgreys**
The focus of this first of two PR is to update the currently existing code that set sample rates. The second PR will aim to add other integrations not currently supporting trace search configuration.

- [x] Update Config API for global and integration configurations
- [x] Update `aiohttp` integration
- [x] Update `bottle` integration
- [x] Update `django` integration
- [x] Update `falcon` integration
- [x] Update `flask` integration
- [x] Update `molten` integration
- [x] Update `pylons` integration
- [x] Update `pyramid` integration
- [x] Update `tornado` integration

Also add non-web integrations:
- [x] aiobotocore
- [ ] aiopg
- [ ] asyncio
- [x] boto
- [x] botocore
- [x] cassandra
- [ ] celery
- [x] dbapi
- [ ] elasticsearch
- [ ] flask_cache
- [ ] futures
- [ ] gevent
- [ ] grpc
- [ ] httplib
- [ ] jinja2
- [ ] kombu
- [ ] mako
- [ ] mongoengine
- [ ] mysql
- [ ] mysqldb
- [x] psycopg
- [ ] pylibmc
- [ ] pymemcache
- [ ] pymongo
- [ ] pymysq
- [x] redis
- [x] rediscluster
- [x] requests
- [ ] sqlalchemy
- [ ] sqlite3
- [ ] vertica

### CODE REVIEW

**Comment by brettlangdon** [comment]
@majorgreys there are some conflicts here with `master`

File: tests/contrib/aiohttp/test_request.py [hunk 1]
```diff
@@ -55,28 +55,123 @@ def test_full_request(self):
...
+                We expect the root span to have the appropriate tag
+        """
         # it should create a root span when there is a handler hit
         # with the proper tags
-        with self.override_config('aiohttp', dict(event_sample_rate=1)):
-            request = yield from self.client.request('GET', '/template/')
-            eq_(200, request.status)
-            yield from request.text()
+        with self.override_global_config(dict(analytics=True)):
+            with self.override_config('aiohttp', dict()):
```
Comment (brettlangdon) [inline tests/contrib/aiohttp/test_request.py]: Why is this here?

File: ddtrace/settings.py [hunk 2]
```diff
@@ -156,6 +161,16 @@ def header_is_traced(self, header_name):
             else self.global_config.header_is_traced(header_name)
         )
 
+    def is_analytics_enabled(self, enabled_if_global=False):
```
[the remark this answers is not in the document]
Comment (majorgreys) [reply ddtrace/settings.py]: Removed this optional argument for now as it's intention was for when we support non-web frameworks. Will revisit this later.

File: ddtrace/contrib/aiohttp/middlewares.py [hunk 3]
```diff
@@ -47,8 +47,9 @@ def attach_context(request):
         )
 
         # Configure trace search sample rate
-        if config.aiohttp.event_sample_rate is not None:
-            request_span.set_tag(EVENT_SAMPLE_RATE_KEY, config.aiohttp.event_sample_rate)
+        analytics_sample_rate = config.aiohttp.get_analytics_sample_rate()
+        if analytics_sample_rate:
```
Comment (brettlangdon) [inline ddtrace/contrib/aiohttp/middlewares.py]: What if the same rate is 0?

File: ddtrace/contrib/bottle/trace.py [hunk 4]
```diff
@@ -1,17 +1,35 @@
...
 
+# Configure default configuration
+config._add('bottle', dict(
+    # Bottle service configuration
+    # DEV: Environment variable 'DATADOG_SERVICE_NAME' used for backwards compatibility
+    service_name=os.environ.get('DATADOG_SERVICE_NAME') or 'bottle',
+    app='bottle',
+    app_type=AppTypes.web,
+
+    distributed_tracing_enabled=False,
```
Comment (brettlangdon) [inline ddtrace/contrib/bottle/trace.py]: Are any of these from here up being used anywhere?

File: ddtrace/contrib/aiohttp/middlewares.py (hunk 3 above)
Comment (majorgreys) [reply ddtrace/contrib/aiohttp/middlewares.py]: If rate is 0, then the metric is not set, so the effect would be the same as disabling analytics.

While the end result of setting it to 0 should be to not have traces consumed by Trace Search, I can see squashing 0 and `None` in this case as bad practice for code clarity and debugging.

File: ddtrace/contrib/bottle/trace.py (hunk 4 above)
Comment (majorgreys) [reply ddtrace/contrib/bottle/trace.py]: No, this was the first integration I tackled, and had hoped to add the config API to each integration, but ultimately that made little sense. Will remove this.

File: ddtrace/contrib/aiohttp/middlewares.py (hunk 3 above)
Comment (rochdev) [reply ddtrace/contrib/aiohttp/middlewares.py]: The result is not the same if it's also configured in the agent.

Comment (brettlangdon) [reply ddtrace/contrib/aiohttp/middlewares.py]: Correct, we need `0` and "not set" to be different values.

If they want to explicitly set as `0` to disable analytics for a given integration then what you have here will not set it meaning it'll use whatever is set in the agent, which may not be `0`.

If that makes sense?

File: ddtrace/contrib/pyramid/patch.py [hunk 5]
```diff
@@ -28,9 +31,13 @@ def traced_init(wrapped, instance, args, kwargs):
     settings = kwargs.pop('settings', {})
     service = os.environ.get('DATADOG_SERVICE_NAME') or 'pyramid'
     distributed_tracing = asbool(get_env('pyramid', 'distributed_tracing', True))
+    analytics = asbool(get_env('pyramid', 'analytics', None))
```
Comment (brettlangdon) [inline ddtrace/contrib/pyramid/patch.py]: Can `asbool()` return `None`?

File: ddtrace/settings/config.py [hunk 6]
```diff
@@ -19,10 +21,17 @@ def __init__(self):
         # use a dict as underlying storing mechanism
         self._config = {}
         self._http = HttpConfig()
+        # Master switch for turning on and off trace search by default
+        self.analytics = asbool(environ.get('DD_ANALYTICS', False))
 
     def __getattr__(self, name):
         if name not in self._config:
             self._config[name] = IntegrationConfig(self)
```
Comment (brettlangdon) [inline ddtrace/settings/config.py]: What about:

```python
self._config[name] = IntegrationConfig(self, name)
```

Then we can handle the `get_env(self.name, 'analytics')` inside the `IntegrationConfig.__init__` ?

File: ddtrace/settings/config.py [hunk 7]
```diff
@@ -67,6 +76,11 @@ def _add(self, integration, settings, merge=True):
         else:
             self._config[integration] = IntegrationConfig(self, settings)
 
+        # Inject environment variables for integration if none set
+        analytics = get_env(integration, 'analytics')
+        if self._config[integration].analytics is None and analytics is not None:
```
Comment (brettlangdon) [inline ddtrace/settings/config.py]: We also might not need this here if we move into `IntegrationConfig.__init__`

File: ddtrace/settings/integration.py [hunk 8]
```diff
@@ -35,9 +35,10 @@ def __init__(self, global_config, *args, **kwargs):
         object.__setattr__(self, 'hooks', Hooks())
         object.__setattr__(self, 'http', HttpConfig())
 
-        # Set default keys/values
+        # Set default analytics configuration, default is disabled
         # DEV: Default to `None` which means do not set this key
-        self['event_sample_rate'] = None
+        self['analytics'] = self.get('analytics', None)
```
Comment (brettlangdon) [inline ddtrace/settings/integration.py]: You should be able to do:

```python
self.setdefault('analytics', None)
self.setdefault('analytics_sample_rate', 1.0)
```

File: ddtrace/contrib/pyramid/patch.py (hunk 5 above)
Comment (majorgreys) [reply ddtrace/contrib/pyramid/patch.py]: No, I need to handle this with an if-else

File: ddtrace/contrib/aiobotocore/patch.py [hunk 9]
```diff
@@ -117,4 +119,8 @@ def _wrapped_api_call(original_func, instance, args, kwargs):
         if request_id2:
             span.set_tag('aws.requestid2', request_id2)
 
+        analytics_sample_rate = config.aiobotocore.get_analytics_sample_rate(use_global_config=False)
+        if analytics_sample_rate:
```
Comment (brettlangdon) [inline ddtrace/contrib/aiobotocore/patch.py]: We still need to support `analytics_sample_rate = 0` here

Comment (brettlangdon) [reply ddtrace/contrib/aiobotocore/patch.py]: We might be able to do something like:


```python
span.set_tag(ANALYTICS_SAMPLE_RATE_KEY, config.aiobotocore.get_analytics_sample_rate(use_global_config=False))`
```

And just have our custom code in `set_tag` determine if the value should be set based on the value returned, e.g. `None` means ignore/do not set, `true` means `1.0`, `false` means `0.0`, and any other float value `0 >= x <= 1.0` gets set.

File: ddtrace/contrib/botocore/patch.py [hunk 10]
```diff
@@ -71,4 +73,8 @@ def patched_api_call(original_func, instance, args, kwargs):
         span.set_tag(http.STATUS_CODE, result['ResponseMetadata']['HTTPStatusCode'])
         span.set_tag("retry_attempts", result['ResponseMetadata']['RetryAttempts'])
 
+        analytics_sample_rate = config.botocore.get_analytics_sample_rate(use_global_config=False)
```
Comment (brettlangdon) [inline ddtrace/contrib/botocore/patch.py]: We should probably reverse the default for `use_global_config`.

Have it be `False` by default and override for the specific integrations we want to inherit from it.

The reason for this is if we ever want to change the default from `False` to `True` (all integrations inherit by default) then we'll have to go through everywhere and update this arg.

Comment (brettlangdon) [reply ddtrace/contrib/botocore/patch.py]: I don't have a strong opinion here, so if you think `True` is the better default then :+1:

File: ddtrace/contrib/pyramid/patch.py [hunk 11]
```diff
@@ -1,7 +1,10 @@
 import os
 
 from .trace import trace_pyramid, DD_TWEEN_NAME
-from .constants import SETTINGS_SERVICE, SETTINGS_DISTRIBUTED_TRACING
+from .constants import (
+    SETTINGS_SERVICE, SETTINGS_DISTRIBUTED_TRACING,
+    SETTINGS_ANALYTICS, SETTINGS_ANALYTICS_SAMPLE_RATE
```
Comment (brettlangdon) [inline ddtrace/contrib/pyramid/patch.py]:
```suggestion
    SETTINGS_ANALYTICS, SETTINGS_ANALYTICS_SAMPLE_RATE,
```

File: ddtrace/contrib/pyramid/patch.py [hunk 12]
```diff
@@ -28,9 +31,15 @@ def traced_init(wrapped, instance, args, kwargs):
     settings = kwargs.pop('settings', {})
     service = os.environ.get('DATADOG_SERVICE_NAME') or 'pyramid'
     distributed_tracing = asbool(get_env('pyramid', 'distributed_tracing', True))
+    # DEV: integration-specific analytics flag can be not set but still enabled
+    # globally for web frameworks
+    analytics = asbool(get_env('pyramid', 'analytics')) if get_env('pyramid', 'analytics') else None
+    analytics_sample_rate = get_env('pyramid', 'analytics_sample_rate', 1.0)
```
Comment (brettlangdon) [inline ddtrace/contrib/pyramid/patch.py]: We don't do this for Django or others do we?

Comment (brettlangdon) [reply ddtrace/contrib/pyramid/patch.py]: Should we?

File: ddtrace/contrib/pyramid/trace.py [hunk 13]
```diff
@@ -61,6 +63,8 @@ def trace_tween_factory(handler, registry):
     tracer = settings.get(SETTINGS_TRACER) or ddtrace.tracer
     enabled = asbool(settings.get(SETTINGS_TRACE_ENABLED, tracer.enabled))
     distributed_tracing = asbool(settings.get(SETTINGS_DISTRIBUTED_TRACING, True))
+    analytics = asbool(settings.get(SETTINGS_ANALYTICS, False)) if settings.get(SETTINGS_ANALYTICS) else None
```
Comment (brettlangdon) [inline ddtrace/contrib/pyramid/trace.py]: Do we need `asbool` here?

Comment (brettlangdon) [reply ddtrace/contrib/pyramid/trace.py]: If `SETTINGS_ANALYTICS` is in `settings`, then calling `settings.get(SETTINGS_ANALYTICS, False)` should never hit the default value.

Can probably reduce this all to:

```python
analytics = settings.get(SETTINGS_ANALYTICS)
```

Default will be `None` which is what you want. You are calling `asbool()` when parsing env variable, so if they choose to set as a non-bool manually, that is on them.

File: ddtrace/contrib/requests/connection.py [hunk 14]
```diff
@@ -72,6 +73,15 @@ def _wrap_send(func, instance, args, kwargs):
         # update the span service name before doing any action
         span.service = _extract_service_name(instance, span, hostname=hostname)
 
+        # Configure trace search sample rate
+        # DEV: Not enabled by default when global analytics config is enabled
+
+        analytics = config.get_from(instance).get('analytics')
```
Comment (brettlangdon) [inline ddtrace/contrib/requests/connection.py]:
```python
cfg = config.get_from(instance)

analytics = cfg.get('analytics')
analytics_sample_rate = cfg.get('analytics_sample_rate')

if cfg.get('distributed_tracing'):
    pass
```

Comment (brettlangdon) [reply ddtrace/contrib/requests/connection.py]: `config.get_from(instance)` is expensive to call multiple times here.

File: ddtrace/contrib/tornado/application.py [hunk 15]
```diff
@@ -19,6 +19,8 @@ def tracer_config(__init__, app, args, kwargs):
         'tracer': ddtrace.tracer,
         'default_service': 'tornado-web',
         'distributed_tracing': True,
+        'analytics': None,
```
Comment (brettlangdon) [inline ddtrace/contrib/tornado/application.py]: Env variables?

File: ddtrace/settings/integration.py [hunk 16]
```diff
@@ -32,12 +33,18 @@ def __init__(self, global_config, *args, **kwargs):
...
         object.__setattr__(self, 'global_config', global_config)
+        object.__setattr__(self, 'integration_name', name)
         object.__setattr__(self, 'hooks', Hooks())
         object.__setattr__(self, 'http', HttpConfig())
 
-        # Set default keys/values
+        # Set default analytics configuration, default is disabled
         # DEV: Default to `None` which means do not set this key
-        self['event_sample_rate'] = None
+        # Inject environment variables for integration, override any set in
```
Comment (brettlangdon) [inline ddtrace/settings/integration.py]: Hmm, maybe I had this in the wrong place before, seems like we might want this before the call to `super().__init__()` ?

Since having here will overwrite any set by the init?

e.g.

```python
ic = IntegrationConfig(config, 'molten', analytics=True)
ic is None  # True
```


Right?

File: ddtrace/settings/integration.py [hunk 17]
```diff
@@ -58,6 +65,22 @@ def header_is_traced(self, header_name):
...
 
+    def _is_analytics_enabled(self, use_global_config):
+        # DEV: analytics flag can be None which should not be taken as
+        # enabled when global flag is disabled
+        if use_global_config and self.global_config.analytics:
+            return self.analytics is not False
+        else:
+            return self.analytics is True
+
+    def get_analytics_sample_rate(self, use_global_config=True):
```
Comment (brettlangdon) [inline ddtrace/settings/integration.py]: Like I mentioned before, this might be better as `, use_global_config=False)`

File: ddtrace/settings/integration.py (hunk 16 above)
Comment (brettlangdon) [reply ddtrace/settings/integration.py]: Or, I commented on this before: https://github.com/DataDog/dd-trace-py/pull/828#discussion_r261326903

Comment (brettlangdon) [reply ddtrace/settings/integration.py]: `.setdefault()` is a nice option here.

File: ddtrace/span.py [hunk 18]
```diff
@@ -130,7 +130,13 @@ def set_tag(self, key, value):
             must be strings (or stringable). If a casting error occurs, it will
             be ignored.
         """
+        # handle None and False values as not setting tag
+        if not value:
+            pass
+
         if key in NUMERIC_TAGS:
+            # handle boolean value as default numeric value
+            value = NUMERIC_TAGS_DEFAULT[key] if value is True else value
```
Comment (brettlangdon) [inline ddtrace/span.py]:
```python
if value is not None:
    return self.set_metric(key, float(value))
```

This should be good enough:

```python
>>> float(True)
1.0
>>> float(False)
0.0
>>> float(1)
1.0
>>> float(0)
0.0
>>> float(0.5)
0.5
>>> float(None)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: float() argument must be a string or a number, not 'NoneType'
```

File: ddtrace/contrib/dbapi/__init__.py [hunk 19]
```diff
@@ -161,6 +170,12 @@ def _trace_method(self, method, name, extra_tags, *args, **kwargs):
             s.set_tags(pin.tags)
             s.set_tags(extra_tags)
 
+            # set analytics sample rate if enabled
+            s.set_tag(
```
Comment (brettlangdon) [inline ddtrace/contrib/dbapi/__init__.py]: hnnnng this is great, so much easier to implement

File: ddtrace/contrib/tornado/application.py [hunk 20]
```diff
@@ -19,8 +19,7 @@ def tracer_config(__init__, app, args, kwargs):
         'tracer': ddtrace.tracer,
         'default_service': 'tornado-web',
         'distributed_tracing': True,
-        'analytics': None,
-        'analytics_sample_rate': 1.0,
```
Comment (brettlangdon) [inline ddtrace/contrib/tornado/application.py]: Is removing this going to cause issues?

File: ddtrace/settings/integration.py (hunk 16 above)
Comment (brettlangdon) [reply ddtrace/settings/integration.py]: This comment still applies.

File: ddtrace/span.py [hunk 21]
```diff
@@ -130,14 +130,10 @@ def set_tag(self, key, value):
...
-        if not value:
-            pass
-
         if key in NUMERIC_TAGS:
-            # handle boolean value as default numeric value
-            value = NUMERIC_TAGS_DEFAULT[key] if value is True else value
-            return self.set_metric(key, value)
+            if value is not None:
+                # DEV: handle boolean values as well
+                return self.set_metric(key, float(value))
```
Comment (brettlangdon) [inline ddtrace/span.py]: we might want:

```python
try:
    self.set_metric(key, float(value))
except (TypeError, ValueError):
    # TODO: Log or something?
    pass

return
```

File: tests/contrib/tornado/test_tornado_web.py [hunk 22]
```diff
@@ -332,7 +332,7 @@ def get_settings(self):
         # distributed_tracing needs to be disabled manually
         return {
             'datadog_trace': {
-                'analytics': True,
+                'analytics_enabled': True,
                 'analytics_sample_rate': 0.5,
```
Comment (brettlangdon) [inline tests/contrib/tornado/test_tornado_web.py]: We should have a test that doesn't set `analytics_sample_rate`, since I think it'll actually raise an exception ?

File: ddtrace/contrib/dbapi/__init__.py [hunk 23]
```diff
@@ -47,6 +50,12 @@ def _trace_method(self, method, name, resource, extra_tags, *args, **kwargs):
...
+                ANALYTICS_SAMPLE_RATE_KEY,
```
Comment (brettlangdon) [inline ddtrace/contrib/dbapi/__init__.py]: We want to make sure this is only enabled on execute/query spans and not `fetch*`, or `rollback`, etc.

Comment (brettlangdon) [reply ddtrace/contrib/dbapi/__init__.py]: If someone enabled the `fetch*` methods, their bill will blow up like crazy

File: ddtrace/contrib/dbapi/__init__.py (hunk 19 above)
Comment (brettlangdon) [inline ddtrace/contrib/dbapi/__init__.py]: We can remove this.

File: ddtrace/settings/integration.py [hunk 24]
```diff
@@ -32,12 +33,17 @@ def __init__(self, global_config, *args, **kwargs):
...
 
-        # Set default keys/values
+        # Set default analytics configuration, default is disabled
         # DEV: Default to `None` which means do not set this key
-        self['event_sample_rate'] = None
+        # Inject environment variables for integration
+        analytics_enabled_env = get_env(name, 'analytics_enabled')
+        if analytics_enabled_env is not None:
+            analytics_enabled_env = asbool(analytics_enabled_env)
+        self.setdefault('analytics_enabled', analytics_enabled_env)
```
Comment (brettlangdon) [inline ddtrace/settings/integration.py]: Don't we also need `get_env(name, 'analytics_sample_rate', default=1.0)` here as well?

Comment (brettlangdon) [reply ddtrace/settings/integration.py]: I see we are handling the `None` case below, but we should support the env variable anyways.

File: ddtrace/span.py [hunk 25]
```diff
@@ -130,9 +130,14 @@ def set_tag(self, key, value):
             must be strings (or stringable). If a casting error occurs, it will
             be ignored.
         """
+
         if key in NUMERIC_TAGS:
-            return self.set_metric(key, value)
+            try:
+                self.set_metric(key, float(value))
+            except (TypeError, ValueError):
+                log.debug("error setting numeric metric {}:{}".format(key, value))
```
Comment (brettlangdon) [inline ddtrace/span.py]:
```suggestion
                log.debug('error setting numeric metric {}:{}'.format(key, value))
```

File: ddtrace/contrib/tornado/application.py (hunk 15 above)
Comment (majorgreys) [reply ddtrace/contrib/tornado/application.py]: At the moment, none of the tracer configurations for `tornado` support environment variables, so I tried to avoid adding new behavior. We should revisit this as we bring the integration into the new Config API.

Comment (brettlangdon) [reply ddtrace/contrib/tornado/application.py]: :+1:

File: ddtrace/contrib/dbapi/__init__.py [hunk 26]
```diff
@@ -50,11 +50,12 @@ def _trace_method(self, method, name, resource, extra_tags, *args, **kwargs):
             s.set_tags(pin.tags)
             s.set_tags(extra_tags)
 
-            # set analytics sample rate if enabled
-            s.set_tag(
-                ANALYTICS_SAMPLE_RATE_KEY,
-                config.dbapi2.get_analytics_sample_rate()
-            )
+            # set analytics sample rate if enabled but only for non-FetchTracedCursor
+            if type(self) is not FetchTracedCursor:
```
Comment (brettlangdon) [inline ddtrace/contrib/dbapi/__init__.py]: We inherit from this class in other integrations, so you'd want, `if not isinstance(self, FetchTracedCursor):`

File: tests/contrib/dbapi/test_unit.py [hunk 27]
```diff
@@ -415,6 +415,41 @@ def method():
         assert span.get_metric('db.rowcount') == 123, 'Row count is set as a metric'
         assert span.get_tag('sql.rows') == '123', 'Row count is set as a tag (for legacy django cursor replacement)'
 
+    def test_fetch_analytics_off(self):
```
Comment (brettlangdon) [inline tests/contrib/dbapi/test_unit.py]: `on`?
